File: sdk/csharp/CaribeFiscal.cs
// CaribeFiscal — SDK C# / .NET (ASP.NET) para la API e-CF de la DGII.
//
// .NET 6+ (usa HttpClient + System.Text.Json). Sin dependencias externas.
//
// Ejemplo:
//   var cf = new CaribeFiscal("https://ecf.omegaerp.do", "ck_live_xxx");
//   var res = await cf.EmitInvoiceAsync(payload);   // payload = objeto/anónimo o Dictionary
//   Console.WriteLine(res.GetProperty("eNCF").GetString());
//
// En ASP.NET regístralo como singleton:
//   builder.Services.AddSingleton(new CaribeFiscal(cfg["Ecf:BaseUrl"], cfg["Ecf:ApiKey"]));

using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Http.Json;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;
using System.Web;

namespace CaribeFiscal
{
    public class CaribeFiscalException : Exception
    {
        public int Status { get; }
        public string? Body { get; }
        public CaribeFiscalException(string message, int status = 0, string? body = null) : base(message)
        {
            Status = status;
            Body = body;
        }
    }

    public class CaribeFiscalClient
    {
        private readonly HttpClient _http;
        private readonly string _baseUrl;

        public CaribeFiscalClient(string baseUrl, string apiKey, HttpClient? http = null)
        {
            _baseUrl = baseUrl.TrimEnd('/');
            _http = http ?? new HttpClient();
            _http.Timeout = TimeSpan.FromSeconds(120);
            _http.DefaultRequestHeaders.Add("X-API-Key", apiKey);
        }

        // ── e-CF ─────────────────────────────────────────────────────────────
        public Task<JsonElement> EmitInvoiceAsync(object payload)        => ReqAsync(HttpMethod.Post, "/api/v1/ecf/submit", payload);
        public Task<JsonElement> GetStatusAsync(string trackId)          => ReqAsync(HttpMethod.Get, $"/api/v1/ecf/status/{Uri.EscapeDataString(trackId)}");
        public Task<JsonElement> GetInvoiceAsync(string encf)            => ReqAsync(HttpMethod.Get, $"/api/v1/ecf/{Uri.EscapeDataString(encf)}");
        public Task<JsonElement> CancelInvoiceAsync(string encf, string motivo) => ReqAsync(HttpMethod.Post, "/api/v1/ecf/cancel", new { encf, motivo });
        public Task<JsonElement> GetQueueAsync()                         => ReqAsync(HttpMethod.Get, "/api/v1/ecf/queue");
        public Task<JsonElement> ListInvoicesAsync(IDictionary<string, string>? filters = null) => ReqAsync(HttpMethod.Get, "/api/v1/ecf/" + Qs(filters));
        public Task<JsonElement> ListInboundAsync(IDictionary<string, string>? filters = null)  => ReqAsync(HttpMethod.Get, "/api/v1/inbound" + Qs(filters));

        public async Task<byte[]> DownloadPdfAsync(string encf)
        {
            using var resp = await _http.GetAsync($"{_baseUrl}/api/v1/ecf/{Uri.EscapeDataString(encf)}/pdf");
            if (!resp.IsSuccessStatusCode)
                throw new CaribeFiscalException($"HTTP {(int)resp.StatusCode}", (int)resp.StatusCode);
            return await resp.Content.ReadAsByteArrayAsync();
        }

        // ── NCF ──────────────────────────────────────────────────────────────
        public Task<JsonElement> GetSequencesAsync()                     => ReqAsync(HttpMethod.Get, "/api/v1/ncf/sequences");
        public Task<JsonElement> NextNcfAsync(string type)              => ReqAsync(HttpMethod.Get, $"/api/v1/ncf/next/{Uri.EscapeDataString(type)}");
        public Task<JsonElement> ConfigureSequenceAsync(object seq)      => ReqAsync(HttpMethod.Post, "/api/v1/ncf/configure", seq);
        public Task<JsonElement> DeactivateSequenceAsync(string type)    => ReqAsync(HttpMethod.Delete, $"/api/v1/ncf/{Uri.EscapeDataString(type)}");

        // ── RNC ──────────────────────────────────────────────────────────────
        public Task<JsonElement> LookupRncAsync(string rnc)             => ReqAsync(HttpMethod.Get, $"/api/v1/rnc/{Uri.EscapeDataString(rnc)}");
        public Task<JsonElement> AutocompleteRncAsync(string q)         => ReqAsync(HttpMethod.Get, "/api/v1/rnc/autocomplete" + Qs(new Dictionary<string, string> { ["q"] = q }));
        public Task<JsonElement> BulkRncAsync(IEnumerable<string> rncs) => ReqAsync(HttpMethod.Post, "/api/v1/rnc/bulk", new { rncs });

        // ── Reportes ───────────────────────────────────────────────────────────
        public Task<JsonElement> Report606Async(string s, string e)    => ReportAsync("606", s, e);
        public Task<JsonElement> Report607Async(string s, string e)    => ReportAsync("607", s, e);
        public Task<JsonElement> Report608Async(string s, string e)    => ReportAsync("608", s, e);
        public Task<JsonElement> ReportIt1Async(string s, string e)    => ReportAsync("it1", s, e);
        public Task<JsonElement> ReportIsrAsync(string s, string e)    => ReportAsync("isr", s, e);
        public Task<JsonElement> ReportIr2Async(object body)            => ReqAsync(HttpMethod.Post, "/api/v1/reports/ir2", body);
        public Task<JsonElement> ReportIr1Async(object body)            => ReqAsync(HttpMethod.Post, "/api/v1/reports/ir1", body);
        public Task<JsonElement> ReportTssAsync(string periodo, object empleados) => ReqAsync(HttpMethod.Post, "/api/v1/reports/tss", new { periodo, empleados });
        private Task<JsonElement> ReportAsync(string tipo, string start, string end) => ReqAsync(HttpMethod.Post, $"/api/v1/reports/{tipo}", new { startDate = start, endDate = end });

        // ── Empresa ──────────────────────────────────────────────────────────
        public Task<JsonElement> GetConfigAsync()                       => ReqAsync(HttpMethod.Get, "/api/v1/company/config");
        public Task<JsonElement> UpdateConfigAsync(object cfg)          => ReqAsync(HttpMethod.Put, "/api/v1/company/config", cfg);

        // ── Núcleo HTTP ────────────────────────────────────────────────────────
        private static string Qs(IDictionary<string, string>? p)
        {
            if (p == null || p.Count == 0) return "";
            var qb = HttpUtility.ParseQueryString(string.Empty);
            foreach (var kv in p)
                if (!string.IsNullOrEmpty(kv.Value)) qb[kv.Key] = kv.Value;
            var s = qb.ToString();
            return string.IsNullOrEmpty(s) ? "" : "?" + s;
        }

        private async Task<JsonElement> ReqAsync(HttpMethod method, string path, object? body = null)
        {
            using var req = new HttpRequestMessage(method, _baseUrl + path);
            if (body != null)
                req.Content = new StringContent(JsonSerializer.Serialize(body), Encoding.UTF8, "application/json");

            using var resp = await _http.SendAsync(req);
            var text = await resp.Content.ReadAsStringAsync();
            JsonElement data = default;
            if (!string.IsNullOrEmpty(text))
            {
                try { data = JsonSerializer.Deserialize<JsonElement>(text); } catch { /* respuesta no-JSON */ }
            }

            if (!resp.IsSuccessStatusCode)
            {
                string msg = $"HTTP {(int)resp.StatusCode}";
                if (data.ValueKind == JsonValueKind.Object && data.TryGetProperty("error", out var e))
                    msg = e.GetString() ?? msg;
                throw new CaribeFiscalException(msg, (int)resp.StatusCode, text);
            }
            return data;
        }
    }
}

File: sdk/node/caribefiscal.ts
/**
 * CaribeFiscal — SDK Node.js / TypeScript para la API e-CF de la DGII.
 *
 * Sin dependencias: usa `fetch` nativo (Node 18+). Para Node <18 instala
 * `undici` o `node-fetch` y asígnalo a globalThis.fetch.
 *
 * Ejemplo:
 *   import { CaribeFiscal } from './caribefiscal'
 *   const cf = new CaribeFiscal('https://ecf.omegaerp.do', 'ck_live_xxx')
 *   const res = await cf.emitInvoice({ ... })
 *   console.log(res.eNCF)
 */

export interface EcfItem {
  numeroLinea: number
  indicadorBienServicio: '1' | '2'   // 1=bien, 2=servicio
  nombre: string
  cantidad: number
  precioUnitario: number
  descuento?: number
  recargo?: number
  tasaITBIS?: 0 | 16 | 18
}

export interface SubmitEcf {
  ecfType: '31'|'32'|'33'|'34'|'41'|'43'|'44'|'45'|'46'|'47'
  emisor: {
    rnc: string; razonSocial: string; direccion: string; fechaEmision: string
    nombreComercial?: string; telefono?: string; email?: string
  }
  comprador?: { rnc?: string; cedula?: string; razonSocial: string; direccion?: string; email?: string }
  items: EcfItem[]
  tipoIngresos: '01'|'02'|'03'|'04'|'05'|'06'
  tipoPago: '1'|'2'|'3'
  formaPago: '01'|'02'|'03'|'04'|'05'|'06'|'07'|'08'|'09'
  invoiceDate: string
  plazoCredito?: string
  fechaVencimientoSecuencia?: string
  informacionReferencia?: {
    ncfModificado: string; fechaNCFModificado: string
    codigoModificacion: '1'|'2'|'3'|'4'|'5'
    rncOtroContribuyente?: string; razonModificacion?: string; indicadorNotaCredito?: '0'|'1'
  }
}

export interface SubmitResult {
  success: boolean; eNCF: string; trackId: string
  status: 'PENDING'|'SUBMITTED'|'ACCEPTED'|'REJECTED'
  invoiceId: string; securityCode: string
}

export class CaribeFiscalError extends Error {
  constructor(message: string, public status: number, public body?: unknown) {
    super(message)
    this.name = 'CaribeFiscalError'
  }
}

export class CaribeFiscal {
  private baseUrl: string
  constructor(baseUrl: string, private apiKey: string, private timeoutMs = 120_000) {
    this.baseUrl = baseUrl.replace(/\/$/, '')
  }

  // ── e-CF ──────────────────────────────────────────────────────────────────
  emitInvoice(payload: SubmitEcf)              { return this.req<SubmitResult>('POST', '/api/v1/ecf/submit', payload) }
  getStatus(trackId: string)                   { return this.req('GET', `/api/v1/ecf/status/${encodeURIComponent(trackId)}`) }
  getInvoice(encf: string)                     { return this.req('GET', `/api/v1/ecf/${encodeURIComponent(encf)}`) }
  cancelInvoice(encf: string, motivo: string)  { return this.req('POST', '/api/v1/ecf/cancel', { encf, motivo }) }
  getQueue()                                   { return this.req('GET', '/api/v1/ecf/queue') }
  listInvoices(filters: Record<string, unknown> = {}) { return this.req('GET', '/api/v1/ecf/' + this.qs(filters)) }
  listInbound(filters: Record<string, unknown> = {})  { return this.req('GET', '/api/v1/inbound' + this.qs(filters)) }

  /** Devuelve el PDF como ArrayBuffer (usa Buffer.from(...) para guardarlo). */
  async downloadPdf(encf: string): Promise<ArrayBuffer> {
    const res = await this.raw('GET', `/api/v1/ecf/${encodeURIComponent(encf)}/pdf`)
    if (!res.ok) throw new CaribeFiscalError(`HTTP ${res.status}`, res.status)
    return res.arrayBuffer()
  }

  // ── NCF ───────────────────────────────────────────────────────────────────
  getSequences()                               { return this.req('GET', '/api/v1/ncf/sequences') }
  nextNcf(type: string)                        { return this.req('GET', `/api/v1/ncf/next/${encodeURIComponent(type)}`) }
  configureSequence(seq: Record<string, unknown>) { return this.req('POST', '/api/v1/ncf/configure', seq) }
  deactivateSequence(type: string)             { return this.req('DELETE', `/api/v1/ncf/${encodeURIComponent(type)}`) }

  // ── RNC ───────────────────────────────────────────────────────────────────
  lookupRnc(rnc: string)                       { return this.req('GET', `/api/v1/rnc/${encodeURIComponent(rnc)}`) }
  autocompleteRnc(q: string)                   { return this.req('GET', '/api/v1/rnc/autocomplete' + this.qs({ q })) }
  bulkRnc(rncs: string[])                      { return this.req('POST', '/api/v1/rnc/bulk', { rncs }) }

  // ── Reportes ────────────────────────────────────────────────────────────────
  report606(startDate: string, endDate: string) { return this.req('POST', '/api/v1/reports/606', { startDate, endDate }) }
  report607(startDate: string, endDate: string) { return this.req('POST', '/api/v1/reports/607', { startDate, endDate }) }
  report608(startDate: string, endDate: string) { return this.req('POST', '/api/v1/reports/608', { startDate, endDate }) }
  reportIt1(startDate: string, endDate: string) { return this.req('POST', '/api/v1/reports/it1', { startDate, endDate }) }
  reportIsr(startDate: string, endDate: string) { return this.req('POST', '/api/v1/reports/isr', { startDate, endDate }) }
  reportIr2(body: Record<string, unknown>)      { return this.req('POST', '/api/v1/reports/ir2', body) }
  reportIr1(body: Record<string, unknown>)      { return this.req('POST', '/api/v1/reports/ir1', body) }
  reportTss(periodo: string, empleados: unknown[]) { return this.req('POST', '/api/v1/reports/tss', { periodo, empleados }) }

  // ── Empresa ─────────────────────────────────────────────────────────────────
  getConfig()                                  { return this.req('GET', '/api/v1/company/config') }
  updateConfig(cfg: Record<string, unknown>)   { return this.req('PUT', '/api/v1/company/config', cfg) }

  // ── Núcleo HTTP ───────────────────────────────────────────────────────────
  private qs(params: Record<string, unknown>): string {
    const entries = Object.entries(params).filter(([, v]) => v != null && v !== '')
    if (!entries.length) return ''
    return '?' + entries.map(([k, v]) => `${encodeURIComponent(k)}=${encodeURIComponent(String(v))}`).join('&')
  }

  private raw(method: string, path: string, body?: unknown): Promise<Response> {
    const ctrl = new AbortController()
    const timer = setTimeout(() => ctrl.abort(), this.timeoutMs)
    const headers: Record<string, string> = { 'X-API-Key': this.apiKey, Accept: 'application/json' }
    if (body !== undefined) headers['Content-Type'] = 'application/json'
    return fetch(this.baseUrl + path, {
      method, headers,
      body: body !== undefined ? JSON.stringify(body) : undefined,
      signal: ctrl.signal,
    }).finally(() => clearTimeout(timer))
  }

  private async req<T = any>(method: string, path: string, body?: unknown): Promise<T> {
    const res = await this.raw(method, path, body)
    const text = await res.text()
    const data = text ? JSON.parse(text) : {}
    if (!res.ok) {
      throw new CaribeFiscalError(data?.error ?? `HTTP ${res.status}`, res.status, data)
    }
    return data as T
  }
}
